pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -1197,6 +1305,85 @@ func (nc *Controller) handleDisruption(zoneToNodeConditions map[string][]*v1.Nod
 	}
 }
 
+func (nc *Controller) podUpdated(oldPod, newPod *v1.Pod) {
+	if newPod == nil {
+		return
+	}
+	if len(newPod.Spec.NodeName) != 0 && (oldPod == nil || newPod.Spec.NodeName != oldPod.Spec.NodeName) {
+		podItem := podUpdateItem{newPod.Namespace, newPod.Name}
+		nc.podUpdateQueue.AddRateLimited(podItem)
+	}
+}
+
+func (nc *Controller) doPodProcessingWorker() {
+	for {
+		obj, shutdown := nc.podUpdateQueue.Get()
+		// "podUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
+			return
+		}
+
+		podItem := obj.(podUpdateItem)
+		nc.processPod(podItem)
+	}
+}
+
+func (nc *Controller) processPod(podItem podUpdateItem) {
```

Comment:
This function requires extensive comment how the logic is handled:
- different scenarios that may happen (e.g. pod observed, node not yet observed; pod observed, node already deleted, etc.) and how each of them is handled

Reply:
Comment added.

Reply:
This comment is a big hard to parse.

// processPod is processing events of assigning pods to nodes. In particular:
// 1. for NodeReady=true node, taint eviction for this pod will be cancelled
// 2. for NodeReady=false or unknown node, taint eviction of pod will happen and pod will be marked as not ready
// 3. if node doesn't exist in cache, it will be skipped and handled later by doEvictionPass


[all other cases are kind of obvious]